emoji: Switch to `zulip_emoji_map.py` for generating emoji info.

HarshitOnGitHub opened the pull request:
Should be merged after PR #7026 gets merged.

HarshitOnGitHub commented:
@timabbott @rishig This is finally ready for review. This contains only one script, other two are WIP for which I will create a separate PR.

HarshitOnGitHub commented:
(As a sidenote, this should be merged after #7026 goes in.)

HarshitOnGitHub commented:
@rishig Updated! :)

rishig commented on tools/setup/emoji/zulip_emoji_map.py:
```diff
@@ -0,0 +1,1505 @@
+from typing import Any, Dict
+
+ZULIP_EMOJI_MAP = {
```
should this be called EMOJI_NAMES?
Potentially this file also should be emoji_names.py?

HarshitOnGitHub replied:
I like `zulip_emoji_map`(lowercase) more since that clearly conveys that it is a mapping not a list.

rishig replied:
hm, possibly I think each entry should indeed just be a list of names :). I haven't read the code that interacts with this, though.

I think in general it's not that useful to put type information in the variable names though, especially now that we have mypy. E.g. I think it'll be pretty clear from the code like the following what is going on:
```
names = emoji_names[codepoint]
// do something with names['canonical_names'], or names['aliases']
```

rishig replied:
This is a big renaming change if we rename it, so we can wait for Tim to comment :).

rishig replied:
Talked to Tim, let's go with `emoji_name_maps` for now. Though I suspect we might actually want to make this emoji_name_lists and just have them be lists :).

rishig commented on tools/setup/emoji/zulip_emoji_map.py:
```diff
@@ -0,0 +1,1505 @@
+    '231b': {'canonical_name': 'times_up', 'aliases': ['hourglass_done']},
+    # seems like the better hourglass. Also see Objects/36
+    '23f3': {'canonical_name': 'time_ticking', 'aliases': ['hourglass']},
+    '1f4e1': {'canonical_name': 'satellite_ante
```
extra space here before the `#`?

HarshitOnGitHub replied:
Sorry didn't get what you meant here?

rishig replied:
hm, not sure how that comment ended up here. I meant in the last line of the file, there seems to be an extra space before the `#`.

HarshitOnGitHub replied:
If that "last line" refers to an explanation then that should have been fixed in this version as I have updated the related script.

rishig replied:
looks like it is fixed, maybe I was looking at an old version. :+1:

rishig commented on tools/setup/emoji/generate_emoji_listing:
```diff
@@ -0,0 +1,219 @@
+        <table>
+            <thead>
+                <tr>
+                    <th class="new-sorting-info">Sorting info</th>
```
Lets make the titles: Category, Codepoint, Images, Zulip, Iamcal (Slack), Gemoji (unordered), Unicode

HarshitOnGitHub replied:
Changed the title except: `Codepoint`. I think `Emoji code` is more appropriate?

rishig replied:
sounds good

rishig commented on tools/setup/emoji/generate_emoji_listing:
```diff
@@ -0,0 +1,219 @@
+
+def get_sorting_info(category, sort_order):
+    # type: (str, int) -> str
+    return ", ".join([category, str(sort_order)])
```
I think no comma here. "People 28" is easier to search for than "People, 28".

HarshitOnGitHub replied:
Done!

rishig commented on tools/setup/emoji/emoji_setup_utils.py:
```diff
@@ -254,63 +244,20 @@ def emoji_is_universal(emoji_dict):
-        codepoint = emoji_dict["unified"].lower()
-        extended_codepoint_to_name[codepoint] = emoji_name
-    return extended_codepoint_to_name
+def generate_name_to_codepoint_map(ZULIP_EMOJI_MAP):
```
I think a python convention is that all caps is used only for constants. So this should either be lowercase (zulip_emoji_map, or even, emoji_names), or not take an argument at all.

There are a few more instances of this in this file, I think

HarshitOnGitHub replied:
I want all of these functions to be pure that way they are easy to debug. Changed `ZULIP_EMOJI_MAP -> zulip_emoji_map`.

HarshitOnGitHub commented on tools/setup/emoji/zulip_emoji_map.py:
```diff
@@ -0,0 +1,1505 @@
+from typing import Any, Dict
```
Done!

HarshitOnGitHub commented:
@rishig Thanks for the careful review. I have also pushed a version of script used to parse excel sheet, can you check out that? Thanks! :)

rishig commented:
For generate_zulip_emoji_map, I think it should take a csv or tsv (tab separated value), rather than an excel file. That will be easier to interact with, debug, etc, and we aren't using any of the fancy features of excel (like formulas). 

For testing this, I think the only test that needs to be done will be:
A = spreadsheet
B = names_dict(A)
C = spreadsheet(B)
D = names_dict(C)

and then check that A == C and B == D.

HarshitOnGitHub commented:
@rishig Updated emoji names import tool to interact with a tsv file. Working on the emoji name tool which I think would be tricky to do, I am assuming we would need explanations in the exported tsv file?

rishig commented:
After talking with Tim, I think we do want csv instead of tsv, since tabs are hard to interact with (e.g. in `git diff`). I realize it's a bit harder to do the parsing, but hopefully something like https://docs.python.org/2/library/csv.html will handle it?

We do need the explanations in the exported file, but I don't understand 
> Working on the emoji name tool which I think would be tricky to do

What do you mean by the emoji name tool, and which part is tricky?

HarshitOnGitHub commented:
> After talking with Tim, I think we do want csv instead of tsv, since tabs are hard to interact with (e.g. in git diff).

I choose tsv because we are using commas in explanation and I suspected we would have to face double quotation on round trip but [csv](https://docs.python.org/3/library/csv.html) seems to have a good support for these scenarios. I have changed the tools to use csv format.

>What do you mean by the emoji name tool, and which part is tricky?

Sorry that was a typo, I was referring to `zulip_emoji_map.py` exporting tool. The good thing is that it is ready.

I have pushed new changes which include:
1: New emoji names export tool.
2: Changed the scripts to use csv format.
3: Modified the emoji names import tool to add `# ignorelongline` comment automatically for line longer than 120.
4: A minor fix in `custom_check.py`.

 :tada: :relaxed:

rishig commented:
> Modified the emoji names import tool to add # ignorelongline comment automatically for line longer than 120.

The main benefit of storing in this format instead of just storing the csv is so that we don't have to check in the long lines into the codebase :). The issue with long lines is it can make the output of `git grep` messy.

lonerz commented:
Seems like zulipbot is going a little crazy here..

timabbott commented:
I merged a couple preparatory commits; it looks like there's still some feedback from @rishig that needs to be addressed.

HarshitOnGitHub commented:
@timabbott Ready for a final review!

HarshitOnGitHub commented:
@timabbott I went over this PR again and I think this is ready for a review from you. Once this gets merged I will be more confident in upgrading the `emoji-datasource` package. :)

HarshitOnGitHub commented:
@timabbott Now that 1.8 is out, what's your plan regarding this?

timabbott commented:
I think we should integrate the upgrade to the emoji-data library that fixes the misplaced emoji, and then rebase this on top of that?  I'm a little scared of this change, so I'd rather tackle this in a couple weeks, whereas that other change seems relatively safe.

HarshitOnGitHub commented:
I have rebased this and added a tool to validate the scripts by doing a round-trip conversion. But it is giving false errors because of presence of stray spaces at the end of explanation lines. We will need to run the excel sheet through some tool/script to remove those spaces. Otherwise this is ready for a review.
@rishig FYI> :)

HarshitOnGitHub commented:
@timabbott This is ready from my and @rishig 's side, so ready for a final sign-off from you. :tada:

timabbott commented:
Merged as the series of commits ending with 811ac718d631a993b02c773b9f2550099f01428b, thanks @HarshitOnGitHub!

timabbott commented:
BTW this was a huge project, so congratulations on getting it finished!
